implement routing and menu/component features for react.js

benbieler opened the pull request:
- [x] routing
- [x] rebase master
- [x] cleanup routing
- [x] create 404 site
- [x] repair node 5.0 stuff
  - [x] fix webpack build
  - [x] repair phpstorm config
  - [x] fix tests
- [x] remove foundation, install react-bootstrap
- [x] css build fails
- [x] fix issues with vagrant
- [x] use mocha as testing engine
- [x] implement menu
- [x] fix head comments
- [x] composer update
- [x] configure frontend translation
  - [x] register component
  - [x] make language configurable through cookies
- [x] `npm test` doesn't execute all tests on linux depending on their path depth
- [x] implement security mock
  - [x] refactor cookie api
  - [x] create simple store and action mock
- [x] create abstract component
  - [x] create simple component that generates a fresh page and has a configurable menu
  - [x] implement adapter to the user system

resolves #11

benbieler committed:
Task #11 created basic mock for the react.js frontend

Ma27 commented:
hmm, the doctrine bundle has a security vulnerability (on Sententiaregum-10 I've installed a version where that vulnerability is fixed).
I should cherry-pick that commit, so you have to rebase master (I'll notify you when you can do that)

Ma27 commented on package.json:
```diff
@@ -2,7 +2,7 @@
   "version": "0.0.0",
   "description": "A social network based on Symfony2 and ReactJS",
-  "main": "index.js",
+  "main": "App.js",
```
should be src/Frontend/App.js, shouldn't it?

Ma27 commented on src/Frontend/components/helloWorld.js:
```diff
@@ -14,5 +14,4 @@ class HelloWorld extends React.Component {
 }
-
-export default HelloWorld;
+export default HelloWorld;
```
please add always a newline at the end of the file

Ma27 commented on src/Frontend/config/routes.js:
```diff
@@ -0,0 +1,11 @@
+var DefaultRoute = Router.DefaultRoute;
+var Route = Router.Route;
+
+module.exports = (
```
ecmascript6 style please (export default instead of module.exports)

Ma27 commented on Gruntfile.js:
```diff
@@ -27,7 +27,7 @@ module.exports = function (grunt) {
         browserify: {
             client: {
-                src: ['src/Frontend/**/*.js', '!src/Frontend/__tests__/*.js'],
+                src: ['src/Frontend/**/*.js', '!src/Frontend/__tests__/**/*.js '],
```
you have to add the PathSpec _src/Frontend/App.js_ at the end of the src list in order to tell browserify to render the App.js

Ma27 commented on src/Frontend/App.js:
```diff
@@ -0,0 +1,7 @@
+import React from 'react';
+import  Router from 'react-router';
```
there's one space too much

Ma27 commented:
@benbieler ok, I've fixed the security issue. Please rebase master

Ma27 commented:
btw the commit message should not be "Frontend ...", but "Task ...". The red label matters ;)

Ma27 commented:
but thanks a lot for the great work :+1:

Ma27 commented:
could you please add some tests in the end??

benbieler commented:
ok

Ma27 commented on src/Frontend/App.js:
```diff
@@ -0,0 +1,7 @@
+import routes from './config/routes';
+
+Router.run(routes, function (Root) {
+    React.render(<Root />, document.getElementById('content'));
```
hmm, should we validate whether this element exist, so we should throw a proper error?

pseudocode:

``` javascript
let element = document.getElementById('content');
if (!element || element === undefined) {
    throw new Error('Cannot find page element');
}
```

Ma27 commented on src/Frontend/config/Routes.js:
```diff
@@ -0,0 +1,24 @@
+import Router from 'react-router';
+
+var DefaultRoute = Router.DefaultRoute;
+var Route = Router.Route;
```
alignment

Ma27 commented on Vagrantfile:
```diff
@@ -2,14 +2,14 @@
   config.vm.provider "virtualbox" do |vb|
     # vb settings
-    vb.name = "Sententiaregum VM"
+    vb.name = "SVM"
```
pls remove this from the commit

Ma27 commented on app/SymfonyRequirements.php:
```diff
@@ -579,8 +579,8 @@ function_exists('simplexml_import_dom'),
             version_compare($installedPhpVersion, '5.4.11', '>='),
-            'When using the logout handler from the Symfony Security Component, you should have at least PHP 5.4.11 due to PHP bug #63379 (as a workaround, you can also set invalidate_session to false in the security logout handler configuration)',
-            'Install PHP 5.4.11 or newer if your project uses the logout handler from the Symfony Security Component.'
+            'When using the logout handler from the Symfony security Component, you should have at least PHP 5.4.11 due to PHP bug #63379 (as a workaround, you can also set invalidate_session to false in the security logout handler configuration)',
```
this, too

Ma27 commented on src/Frontend/config/MenuConfig.js:
```diff
@@ -0,0 +1,21 @@
+
+import MenuItem from '../menu/model/MenuItem';
+let config = [
+    new MenuItem('/#!/', 'Label', 'User')
```
the _/#!/_ prefix should be added automatically by the MenuItem class

Ma27 commented on src/Frontend/menu/model/MenuItem.js:
```diff
@@ -0,0 +1,24 @@
+    permissionRule;
+
+    constructor(label, url, permissionRule) {
+        this.label = label;
```
alignment

Ma27 commented on src/Frontend/menu/view/Menu.js:
```diff
@@ -0,0 +1,47 @@
+ * @author Benjamin Bieler <benjaminbieler2014@gmail.com>
+ */
+class Menu extends React.Component {
+    render() {
```
could you please add some jsdoc stuff to your methods?

Ma27 commented on src/Frontend/security/Security.js:
```diff
@@ -0,0 +1,16 @@
+/**
+ * @author Benjamin Bieler <benjaminbieler2014@gmail.com>
+ */
+class security {
```
Should be uppercase I guess ;)

Ma27 replied:
hmm, I would refactor the whole security system using reflux:
we need a UserStore that is able to get the securityexpression from the menuconfig in order to evaluate it. After that an event will be triggered using the reflux dispatcher (we DON'T return true because this won't work when using ajax requests in order to check roles for example).
The event should be a class from the Frontend/event directory and should contain information about the user and whether he's allowed to see the menu item.
In your case you just need to build the basic class and trigger **always** an event that tells the menu that the user is allowed to see every menuitem

Ma27 commented on Vagrantfile:
```diff
@@ -2,7 +2,7 @@
 
 Vagrant.configure(2) do |config|
-  config.vm.box = "trusty64"
+  config.vm.box = "ubuntu/trusty64"
```
this should be reverted, too as this will be fixed in another PR. (I guess, your machine works fine, so don't need the vagrantfile anymore)

Ma27 commented on src/Frontend/config/Routes.js:
```diff
@@ -0,0 +1,24 @@
+ * Please check out the license file in the document root of this application
+ */
+
+/**
```
jsdoc should be used at classes only

Ma27 commented on src/Frontend/menu/view/Menu.js:
```diff
@@ -0,0 +1,47 @@
+                );
+            }
+        }
+        return (
```
some spaces would be nice, this looks very squeezed ;)

Ma27 commented on src/Frontend/menu/model/MenuItem.js:
```diff
@@ -0,0 +1,24 @@
+/**
+ * @author Benjamin Bieler <benjaminbieler2014@gmail.com>
+ */
+class MenuConfig {
```
this should be moved to a global model directory.
*Frontend/model/menu/MenuConfig.js*

Ma27 commented on src/Frontend/menu/view/Menu.js:
```diff
@@ -0,0 +1,47 @@
+/**
+ * @author Benjamin Bieler <benjaminbieler2014@gmail.com>
+ */
+class Menu extends React.Component {
```
should be moved to *Frontend/components/menu/Menu.js*

Ma27 commented:
rebased this

Ma27 commented:
resolved all conflicts with master, so we can continue working on that branch @benbieler

Ma27 committed:
Task #11 finished frontend setup

- webpack instance with es6 and less
- babel 6
- security system with adapters for the menu and basic component visibility features
- abstract react.js components with security config
- menu with configurable and translatable items
- mocha instead of karma-jasmine
- code style fix (2 instead of 4 spaces in javascript)
- refactored eslint config and introduced less linter
- resolves #11

Ma27 commented:
it seems as node :shit: 5 has some issues with our build :/

Ma27 commented:
ok it was some weird configuration issue. Should do now.

Ma27 commented:
ah karma does not work because of those weird peer deps

Ma27 commented on src/Frontend/styles/custom.scss:
```diff
@@ -8,7 +8,7 @@
  */
 
-@import '../../../node_modules/zurb-foundation-5/scss/foundation';
+$icon-font-path:'~bootstrap-sass/assets/fonts/bootstrap/';
```
obsolete

Ma27 commented on webpack.factory.js:
```diff
@@ -11,28 +11,47 @@
 
-    cssQueryString = 'css-loader?' + cssQueryString + '!sass-loader?outputStyle=expanded&sourceMap';
+    scssQueryString = 'css-loader?' + useMinimizeAndSourceMap + '!sass-loader?outputStyle=expanded&sourceMap';
+    cssQueryString  = 'css-loader?' + useMinimizeAndSourceMap;
```
can be done better

Ma27 commented:
ok now we're using less instead of scss in order to avoid bigger issues with executables

Ma27 commented on src/AppBundle/Composer/ScriptHandler.php:
```diff
@@ -38,6 +38,14 @@ public static function installNpmDependencies(CommandEvent $event)
+        $npm     = (new ExecutableFinder())->find('npm');
+        $handler = function ($type, $buffer) use ($event) {
+            $event->getIO()->write($buffer, false);
+        };
```
should be refactored

Ma27 commented:
can be useful for translations: https://github.com/martinandert/react-translate-component

Ma27 commented on src/Frontend/components/app/Menu.js:
```diff
@@ -0,0 +1,47 @@
+     * @returns {XML}
+     */
+    render() {
+        let items = this.props.items.map(function (item) {
```
es6 style pls

Ma27 commented:
/cc @benbieler 
I guess it is better to use mocha:
sice mocha works properly with the nodejs core, so we don't have any problems with a non-working debugger and big sourcemap issues. Another reason is that mocha supports real commonjs support and since all of the view is abstracted in ReactJS components (there's an integration of reactjs and its testutils into mocha) and the rest is pure business logic, the switch to a node-based engine (karma was browser-based) shouldn't be a hard job
